### PR TITLE
use webgl2 when available

### PR DESCRIPTION
**Pull request by rotu**
Use WebGL2 rendering context when available.

## Launch Checklist

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [X] Briefly describe the changes in this PR.
 - [X] Link to related issues.
 - [X] Include before/after visuals or gifs if this PR includes visual changes.
 - [x] Write tests for all new functionality.
 - [X] Document any changes to public APIs.
 - [x] Post benchmark scores.
 - [x] Manually test the debug page.
 - [X] Add an entry to `CHANGELOG.md` under the `## main` section.

### CODE REVIEW

**Comment by rotu** [comment]
@JannikGM The reason this came up is that I'm using a [custom 3D layer](https://maplibre.org/maplibre-gl-js-docs/example/add-3d-model/) and some shader features that were working in plain threejs (which defaults to webgl2) were failing in my MapLibre layer and vice versa.

File: CHANGELOG.md [hunk 1]
```diff
@@ -12,6 +13,7 @@
 
 ### 🐞 Bug fixes
 - *...Add new stuff here...*
+- Remove dependency on `@rollup/plugin-json`, which was in conflict with `rollup-plugin-import-assert`
```
[the remark this answers is not in the document]
Comment (rotu) [reply CHANGELOG.md:16]: Not accidental. The `@rollup/plugin-json` plugin crashes out when trying to build due to bad interaction with `rollup-plugin-import-assert`: https://github.com/rollup/plugins/pull/1361

This PR is still a draft until I can get maplibre-gl-js to build cleanly.

File: src/ui/map.ts [hunk 2]
```diff
@@ -2557,8 +2557,9 @@ class Map extends Camera {
             }
         }, {once: true});
 
-        const gl = this._canvas.getContext('webgl', attributes) ||
-            this._canvas.getContext('experimental-webgl', attributes);
```
[the remark this answers is not in the document]
Comment (rotu) [reply src/ui/map.ts:2561]: https://caniuse.com/webgl
IE11 is the only browser using `experimental-webgl`, instead of `webgl`. If it's supported my maplibre-gl-js, I'm not sure how to run tests on that platform.

[the remark this answers is not in the document]
Comment (HarelM) [reply src/ui/map.ts:2561]: IE is not supported as of version 2.x

**Comment by dangerden** [comment]
I would also suggest using WebGL2 by default. It might be needed for VAOs available in WebGL2 only and easy state switching between several custom layers.

**Comment by rotu** [comment]
@JannikGM Could you give this a quick look? How do I tell whether users of the WebGL timing functions are getting the right results?
Also, does the heatmap debug page actually work? I'm getting "Cannot add layer "heatmap" before non-existing layer "waterway-label"."

**Comment by klokan** [comment]
Would be amazing to take this into v3 release. People from DeckGL wanted to switch to WebGL2 as well - FYI @HarelM

**Comment by HarelM** [comment]
@klokan I'm watching this PR closely, but a) it's still in draft b) it incorporates a lot of small fixes that should not be part of this PR.
but since it's still in draft I prefer not to start a review and allow the person who opened the PR to let us know when it's ready for review. 😀

**Comment by HarelM** [comment]
Please make sure to add the relevant unit tests to the added functionality.

File: src/gl/timing.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+export type {Timing};
+
+export function getTimingAPI(gl: WebGLRenderingContext) : null | Timing {
+    const result = {} as Partial<Timing>;
+
+    if (isWebGL2(gl)) {
+        const ext = gl.getExtension('EXT_disjoint_timer_query_webgl2');
+        if (!ext) return null;
+        for (const m of [
```
Comment (HarelM) [inline src/gl/timing.ts:37]: This form of binding doesn't allow proper "find all references".
I would advise to simply bind all the methods one after the other without using a for loop on strings.

Comment (rotu) [reply src/gl/timing.ts]: good call

**Comment by HarelM** [comment]
There are changes in `context.ts` and `timing.ts` and some changing in other files, these changes need to be covered in tests. This is a functionality that was added.

**Comment by rotu** [comment]
Can you help me write a test to check that the `Context.timing` function returns something when Context is constructed against a canvas that supports WebGL2? I’m struggling greatly with the testing infrastructure in this repo. I’m happy to flesh out text coverage once I get over that hump.

**Comment by HarelM** [comment]
Check out the info here:
https://github.com/maplibre/maplibre-gl-js/tree/main/test#writing-unit-tests
Basically, you should create a test file that matches the file you changed and use jest to create a unit test around the logic you changed/added.
There are a lot of unit tests in this project, and you can easily get inspiration from them.
It's less about getting the actual webgl2 functionality, although this is important too.
You can test the webgl2 in the integration tests (test like browser tests I guess) and not in the unit tests.
I would focus on adding unit tests to the added functionality and think of a way to test the webgl2 added code, I'm not sure how to do that, maybe the render test can do that.
In any case, this repo is filled with docs on how to write and run the tests.
If the docs are not clear, feel free to submit a PR to improve the docs.

**Comment by rotu** [comment]
@HarelM Every single unit test around `Context` requires using the `headless-gl` library, which is DoA on MacOS, doesn't support WebGL2, and doesn't even mock anything. I don't want to write these tests until I feel confident this works with an actual webgl2 instance.

That `Timing` class is only to facilitate the timing queries in `Map._render` and `Painter.gpuTimingStart`. What existing tests exist around this? How do I check that the timing features work today and continue working after this PR?

**Comment by HarelM** [comment]
Unfortunately, I don't know the answer of which tests exists to cover the classes you changed.
In terms of how you test the functionality you added - the answer is always the same - you know what you changed, and you should know how to test it - even if you know how to manually test it, that a good start, and we can think how to automate it.
If you need guidance in order to use some testing infra or you need to create new kind of tests, feel free to ask. :-)

**Comment by Pessimistress** [comment]
I can confirm that deck.gl defaults to WebGL2, and some features (transition and GPU aggregation) are only available in WebGL2.

Regarding testing: I do not think it is possible to test WebGL2 code paths under node with the existing harness. The [test/integration/browser](https://github.com/maplibre/maplibre-gl-js/tree/main/test/integration/browser) tests are run in (headless) browsers, though at a glance the coverage is limited. If there is either a public option or an internal flag to force create the context in a specific version, we can run the browser tests to check both WebGL1 and 2 support.

**Comment by rotu** [comment]
> you know what you changed, and you should know how to test it

@HarelM Unfortunately that is not the case here. My goal was to keep GPU timing calls functionally unchanged. But I can't figure out how the `gpu-timing-layer` stuff in `Map` and `gpuTiming...` stuff in `Painter` is supposed to work so I can't verify that it is indeed still working. I'll rip this out and see what breaks...

**Comment by HarelM** [comment]
Sure, manual testing is the first phase :-) then we need to think how to automate them.

File: src/render/vertex_array_object.ts [hunk 4]
```diff
@@ -90,39 +90,24 @@ class VertexArrayObject {
...
-        } else {
-            numPrevAttributes = context.currentNumAttributes || 0;
-
-            // Disable all attributes from the previous program that aren't used in
-            // the new program. Note: attribute indices are *not* program specific!
-            for (let i = numNextAttributes; i < numPrevAttributes; i++) {
-                // WebGL breaks if you disable attribute 0, so if i == 0.
-                // http://stackoverflow.com/questions/20305231
-                gl.disableVertexAttribArray(i);
-            }
```
[the remark this answers is not in the document]
Comment (HarelM) [reply src/render/vertex_array_object.ts:124]: We do not support IE 11, I also think there are some checks in the code that we can omit (for example I see a check for supporting webp format, which all modern browsers support nowadays) and write a disclaimer about not supporting old browsers in general.

File: src/gl/webgl2.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+/// <reference lib="dom" />
+
+export function isWebGL2(
+    gl: WebGLRenderingContext
+): gl is WebGL2RenderingContext {
+    return gl.getParameter(gl.VERSION).startsWith('WebGL 2.0');
```
[the remark this answers is not in the document]
Comment (rotu) [reply src/gl/webgl2.ts:6]: This is the standards-compliant way to check whether a given context is WebGL1 or 2.
https://registry.khronos.org/webgl/specs/latest/2.0/#3.7.2

I don’t believe this is slow either - what make you think so and how could I confirm that?

[the remark this answers is not in the document]
Comment (rotu) [reply src/gl/webgl2.ts]: And I had to go back to the previous implementation, but I added memoization. Turns out `canvas` can legally be null (and indeed it is in our unit testing rig)

https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/canvas

File: src/ui/map.ts [hunk 6]
```diff
@@ -2753,33 +2745,6 @@ class Map extends Camera {
...
-                const renderGPUTime = extTimerQuery.getQueryObjectEXT(gpuTimer, extTimerQuery.QUERY_RESULT_EXT) / (1000 * 1000);
-                extTimerQuery.deleteQueryEXT(gpuTimer);
-                this.fire(new Event('gpu-timing-frame', {
-                    cpuTime: renderCPUTime,
-                    gpuTime: renderGPUTime
-                }));
-            }, 50); // Wait 50ms to give time for all GPU calls to finish before querying
-        }
-
-        if (this.listens('gpu-timing-layer')) {
```
[the remark this answers is not in the document]
Comment (rotu) [reply src/ui/map.ts:2769]: Thanks! I tried it out, and this does seem to provide some numbers on chrome. It also causes the map to fail to render on Firefox. For now, I think it's not a very useful feature but I'm open to reintroducing it later down the line!

File: src/gl/context.ts [hunk 7]
```diff
@@ -117,7 +140,6 @@ class Context {
             this.extRenderToTextureHalfFloat = gl.getExtension('EXT_color_buffer_half_float');
         }
 
-        this.extTimerQuery = gl.getExtension('EXT_disjoint_timer_query');
```
[the remark this answers is not in the document]
Comment (rotu) [reply src/gl/context.ts:120]: Removed this because the performance code crashes Firefox anyway. If we find this to be necessary, we can re-add it in the future.

**Comment by rotu** [comment]
I don't know how to interpret the benchmark scores but here they are: [all.pdf](https://github.com/maplibre/maplibre-gl-js/files/10297378/all.pdf)

File: src/gl/webgl2.ts [hunk 8]
```diff
@@ -0,0 +1,12 @@
+const cache = new WeakMap();
```
Comment (HarelM) [inline src/gl/webgl2.ts:1]: I think this can be simplified to a "regular" boolean...?

File: src/render/painter.ts [hunk 9]
```diff
@@ -504,52 +500,7 @@ class Painter {
         if (layer.type !== 'background' && layer.type !== 'custom' && !(coords || []).length) return;
         this.id = layer.id;
 
-        this.gpuTimingStart(layer);
         draw[layer.type](painter, sourceCache, layer as any, coords, this.style.placement.variableOffsets);
-        this.gpuTimingEnd();
-    }
-
-    gpuTimingStart(layer: StyleLayer) {
```
Comment (HarelM) [inline src/render/painter.ts:512]: Why was this removed?

File: src/gl/value.ts [hunk 10]
```diff
@@ -417,19 +418,18 @@ export class BindElementBuffer extends BaseValue<WebGLBuffer> {
...
-    set(v: any) {
-        if (!this.vao || v === this.current && !this.dirty) return;
-        this.vao.bindVertexArrayOES(v);
+    set(v: WebGLVertexArrayObject | null) {
+        if (v === this.current && !this.dirty) return;
+        const gl = this.gl;
+        if (isWebGL2(gl)) {
+            gl.bindVertexArray(v);
+        } else {
+            gl.getExtension('OES_vertex_array_object')?.bindVertexArrayOES(v);
```
Comment (HarelM) [inline src/gl/value.ts:431]: isn't `getExtension` expensive? Can't we cache its response?

File: src/render/painter.ts (hunk 9 above)
Comment (HarelM) [reply src/render/painter.ts:512]: I see this was asked in a different thread.

File: src/gl/value.ts (hunk 10 above)
Comment (rotu) [reply src/gl/value.ts:431]: Unlike `getParameter`, based on profiling in Chrome, I did not find this call to be costly.

Also, if the context is lost and regained, the extension is no longer valid. https://registry.khronos.org/webgl/specs/latest/1.0/#5.14.14

How is *not* a problem for `extVertexArrayObject` and `extTextureFilterAnisotropic` today?

File: src/gl/webgl2.ts (hunk 8 above)
Comment (rotu) [reply src/gl/webgl2.ts:1]: A boolean that lives where? I wanted to keep this as a separate function from `Context` so that it can be used for type narrowing:

```
if (isWebGL2(gl)){
  return gl.createVertexArray(); // Typescript will infer here that it's a WebGL2RenderingContext and won't complain that the method might not exist
}
```

Comment (HarelM) [reply src/gl/webgl2.ts:1]: In the same file, instead of `cache = new WeakMap()` write `let cache = undefined`, and later on check for `=== undefined` and cache the result.

Comment (rotu) [reply src/gl/webgl2.ts:1]: > In the same file, instead of `cache = new WeakMap()` write `let cache = undefined`, and later on check for `=== undefined` and cache the result.

That would be very wrong. A page can have multiple webgl contexts and they don't all need to be the same type. It might even make sense to have a separate contexts for different layers!

File: src/gl/value.ts (hunk 10 above)
Comment (HarelM) [reply src/gl/value.ts:431]: Not sure I understand the last line in your comment, It's that this call was "cached" before in the Context's constructor and now it called every creation of vertex array, it might be negligible, but if done many time it might cause a performance hit. My 2 cents I guess...

File: src/gl/webgl2.ts (hunk 8 above)
Comment (HarelM) [reply src/gl/webgl2.ts:1]: I see, this map is per context, missed that when reading it earlier. Thanks for the hint.

File: src/gl/value.ts (hunk 10 above)
Comment (rotu) [reply src/gl/value.ts:431]: >  Not sure I understand the last line in your comment

I was referring to the spec's behavior on a `Context Lost` event.

When this happens, enabled extensions are disabled and trying to reuse them will cause errors. It seems like, if the rendering context is ever regained, our `Context` object is still holding on to these invalid extensions objects

> Once an extension is enabled, it is only disabled if the WebGL rendering context is lost (see below), with the exception of the "WEBGL_lose_context" extension which remains active through any loss of context. Any objects referenced by a disabled extension, such as the object returned by getExtension, are no longer associated with the WebGL rendering context. Any extension objects that derive from WebGLObject have their [invalidated](https://registry.khronos.org/webgl/specs/latest/1.0/#webgl-object-invalidated-flag) flag set to true. Behavior of extensions' methods after context loss is defined by the steps in the section ["The WebGL context"](https://registry.khronos.org/webgl/specs/latest/1.0/#WEBGLRENDERINGCONTEXT).

Comment (HarelM) [reply src/gl/value.ts:431]: Was this the only place where this was a problem? I mean there are other places that call `getExtension`, this is addressed in this PR as well - so that in case of context lost (assuming we can recover from it) the other places will work as expected?

Comment (rotu) [reply src/gl/value.ts:431]: I would assume it's a problem other places as well. I'm certainly more confident in the code this way. I'm also pretty confident that it's not a performance bottleneck at all. If this *is* a performance issues, [where would I see it?](https://github.com/maplibre/maplibre-gl-js/pull/1891#issuecomment-1364344763)

Comment (HarelM) [reply src/gl/value.ts:431]: We have an issue with the performance check, you're not the only one who has a hard time reading the results...

Comment (rotu) [reply src/gl/value.ts:431]: Another reason I did it this way is so that Value would not need to maintain a reference to the `Context` object, just the standard `WebGLRendering Context`. I get code smells from the `Context`/`IValue` wrapper code, and this seemed the lesser of two evils.

I'd be happy to revisit this in the future once I figure out how to actually measure the performance of this code IRL.

Comment (HarelM) [reply src/gl/value.ts:431]: Fair enough.

**Comment by HarelM** [comment]
One last thing, what tests are there to cover this new ability?
I saw that the render tests only use webgl 1...

**Comment by rotu** [comment]
> One last thing, what tests are there to cover this new ability?
> I saw that the render tests only use webgl 1...

Unfortunately the render tests don’t even use a browser-based implementation of WebGL. So I think that making those render tests run on FireFox and Chrome would be a really great addition to coverage.

I looked at the debug page, manually ran the heatmap layer code, and loaded it into my work project that needed to use webgl2 in a custom layer. Also nothing looks egregiously wrong in the performance logs.

**Comment by HarelM** [comment]
Checkout the CJK browser test, although not covering a lot, but for basic sanity this should do.

File: src/gl/webgl2.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (rotu) [reply src/gl/webgl2.ts:1]: It’s simple memoization. But I think an instanceof check would do so I’ll switch to that.

File: src/gl/webgl2.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (rotu) [reply src/gl/webgl2.ts]: > This is doing string operations and querying state from the driver. Both of which are fast, but still very slow compared to pure just sending data to the driver

I understand that it *can* be slow and I believe your intuition. How can I *see for myself* that it’s slow and dragging down performance? The best I’ve managed are unrealistic microbenchmarks.

> This is only ever necessary if you haven't created the context yourself.

> It's not possible to request a WebGL 1.x context and end-up with a WebGL 2.x context or vice-versa.

That’s true, but we have code (especially in all the render tests) that uses a `gl` object not created via `getContext`.

This brings me to a big headscratcher - what is the purpose of the Context class in the first place? If it is mostly to track the state of the WebGlRenderingContext (so we don’t issue unnecessary API calls), how do we account for changes made in Custom Layers, which have direct access to the WebGL context but not the Context object.

File: src/gl/webgl2.ts [hunk 11]
```diff
@@ -0,0 +1,5 @@
+export function isWebGL2(
+    gl: WebGLRenderingContext
+): gl is WebGL2RenderingContext {
+    return gl.constructor.name === 'WebGL2RenderingContext';
```
[the remark this answers is not in the document]
Comment (rotu) [reply src/gl/webgl2.ts:4]: > I still think this is too complicated and potentially slow for something that can be called thousands of time per frame.

How can I confirm/disconfirm this? I sincerely doubt this string comparison is every going to be a bottleneck. If it is, we could certainly cache the return value of this function in the `Context` constructor.

We can't rely on there being a `Context` object - indeed the `CustomLayerInterface` is [unaware of it](https://github.com/maplibre/maplibre-gl-js/blob/main/src/style/style_layer/custom_style_layer.ts). The `WebGL[2]RenderingContext` is the source of truth, and that's why this is a function *not* relying on some wrapper object.

File: src/ui/map.ts [hunk 12]
```diff
@@ -2643,8 +2643,9 @@ class Map extends Camera {
             }
         }, {once: true});
 
-        const gl = this._canvas.getContext('webgl', attributes) ||
-            this._canvas.getContext('experimental-webgl', attributes);
+        const gl =
+            this._canvas.getContext('webgl2', attributes) ||
```
[the remark this answers is not in the document]
Comment (rotu) [reply src/ui/map.ts:2647]: 1. https://github.com/maplibre/maplibre-gl-js/blob/main/CONTRIBUTING.md#do-not-violate-mapbox-copyright
2. This PR is targeting 3.0, a breaking change version, and I find that old Mapbox extensions tend not to work with MapLibre anyway. It's always possible to add a "always webgl1" option later should this prove necessary.

**Comment by HarelM** [comment]
@rotu where are we on this PR? What's missing in order to merge this?

**Comment by birkskyum** [comment]
I've been exploring moving our tests to WebGL2, which requires us to swap JSDOM and all the browser mocking with a headless browser like puppeteer or playwright, because there is no maintained node implementations of WebGL2 we can use.

For the longest time WebGL2 wasn't widely supported, due to lacking safari support, but it's finally getting there (currently 93.96%, but growing as iPhones get updated), so what I wonder is, could we simply move to webgl2 entirely as the support becomes sufficient, and thus save all this conditional logic and all the gl1/gl2 checks? Would there be any downsides to that?

**Comment by HarelM** [comment]
The downside would be the support of old devices and old browsers.
Of course, one could use an old version of maplibre to support old browsers, but this is the same as what we did with the removal of the support of IE. I'm not saying we shouldn't, we just need to make sure that do that with proper communication.
I would advise to try and concentrate the logic around webgl 1 in a single file as much as possible so that removing support for it would be mainly remove that file.

**Comment by birkskyum** [comment]
@HarelM it's a good point to try to keep the logic of each graphics backend isolated, also with webgpu on the horizon, somewhat similar to the modularization effort on Native. I'd love to see some benchmarks of this PR, as that would be the main selling point of introducing webgl 2. As long as we stay compatible with webgl 1, we can't use many of the optimizations from webgl 2, because of non-backwards compatible shader changes, so it would be interesting to see what the results are without those. Here is the PR from MapLibre Native that has most of the webgl 2 shader changes: https://github.com/maplibre/maplibre-gl-native/pull/995/files#diff-524519fde8010b2a448977ac84de5feefa300da4983495a0a2bbdc30a62d2bd8

**Comment by birkskyum** [comment]
If almost everybody has webgl2 and this opts for that when available, I'm concerned about merging this before having the ability to test webgl2.

**Comment by birkskyum** [comment]
The changes to map.ts has to be resolved for this to proceed. @rotu , would it be possible for you to have a look at the merge conflict? We are getting closer to having some webgl2 tests in place, but I can't tests it against this PR due to the conflicts in map.ts.

**Comment by birkskyum** [comment]
@rotu, the render test suite is now webgl2 compatible. Also the github test workflows has been refactored so it's broken into separate workflows so it's easier to dive into. It does require rebasing these changes to main though.

**Comment by birkskyum** [comment]
@pramilk , I think the #2447 did something here - can you give us some pointers on how to update the code so fadeDuration can be found again in the build step?

File: src/ui/map.ts [hunk 13]
```diff
@@ -2843,15 +2844,6 @@ class Map extends Camera {
      * @private
      */
     _render(paintStartTimeStamp: number) {
-        let gpuTimer, frameStartTime = 0;
-        const extTimerQuery = this.painter.context.extTimerQuery;
-        const fadeDuration = this._idleTriggered ? this._fadeDuration : 0;
```
Comment (birkskyum) [inline src/ui/map.ts:2848]: @rotu , I think the build is breaking, because this is removed.

**Comment by birkskyum** [comment]
Here are benchmarks, using webgl2.
Screenshot 2023-05-12 at 00 14 40

File: src/render/painter.ts [hunk 14]
```diff
@@ -122,7 +121,7 @@ class Painter {
     cache: {[_: string]: Program<any>};
     crossTileSymbolIndex: CrossTileSymbolIndex;
     symbolFadeChange: number;
-    gpuTimers: {[_: string]: any};
+    emptyTexture: Texture;
```
Comment (HarelM) [inline src/render/painter.ts:124]: Is this used? I don't see a code change with this parameter in this PR...

**Review by HarelM — APPROVED** [review]
Overall approved, Thanks for all the hard work.
I left a very small comment about a variable that I'm not sure is being used.
If it is being used feel free to merge, if not, please remove it and merge.

File: src/ui/map.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (birkskyum) [reply src/ui/map.ts:2958]: How is this related to webgl/2

File: src/ui/map.ts [hunk 15]
```diff
@@ -2730,8 +2730,9 @@ class Map extends Camera {
             }
         }, {once: true});
 
-        const gl = this._canvas.getContext('webgl', attributes) ||
-            this._canvas.getContext('experimental-webgl', attributes);
+        const gl =
```
Comment (birkskyum) [inline src/ui/map.ts:2733]: Maybe this could even have typings like

```ts
this._canvas.getContext('webgl2', attributes) as WebGL2RenderingContext ||
            this._canvas.getContext('webgl', attributes) as WebGLRenderingContext;
```

File: src/ui/map.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (HarelM) [reply src/ui/map.ts:2958]: This is debug/testing code, it should be in production. If you want to debug something, create a branch and add this...
We can move this to a readme if anyone is interested...

[the remark this answers is not in the document]
Comment (birkskyum) [reply src/ui/map.ts:2958]: I assume you mean "shouldn't be".

[the remark this answers is not in the document]
Comment (HarelM) [reply src/ui/map.ts:2958]: Yes, thanks.

**Comment by birkskyum** [comment]
I'll make a follow up PR to simplify the type narrowing / add types to the getContext().

File: src/gl/webgl2.ts [hunk 16]
```diff
@@ -0,0 +1,12 @@
+const cache = new WeakMap();
+export function isWebGL2(
+    gl: WebGLRenderingContext
```
[the remark this answers is not in the document]
Comment (birkskyum) [reply src/gl/webgl2.ts:3]: I've added more typings here #2508

File: src/ui/map.ts (hunk 15 above)
Comment (birkskyum) [reply src/ui/map.ts:2733]: I've added more typings here #2508

File: src/render/painter.ts (hunk 14 above)
Comment (birkskyum) [reply src/render/painter.ts:124]: Removed here #2508

File: src/gl/webgl2.ts (hunk 8 above)
Comment (birkskyum) [reply src/gl/webgl2.ts:1]: If you have a better way to do this, feel free to make a PR of this.
